src/main/java/com/flagsmith/flagengine/segments/SegmentEvaluator.java

```diff
@@ -1,196 +1,201 @@
+        return contextValue != null;
 
-  /**
-   * Matches condition value with the trait value.
-   *
-   * @param condition Condition to evaluate with.
-   * @param value Trait value to compare with.
-   */
-  public static Boolean conditionMatchesTraitValue(SegmentConditionModel condition, Object value) {
-    SegmentConditions operator = condition.getOperator();
-    switch (operator) {
-      case NOT_CONTAINS:
-        return (String.valueOf(value)).indexOf(condition.getValue()) == -1;
       case CONTAINS:
-        return (String.valueOf(value)).indexOf(condition.getValue()) > -1;
-      case IN:
-        if (value instanceof String) {
-          return Arrays.asList(condition.getValue().split(",")).contains(value);
-        }
-        if (value instanceof Integer) {
-          return Arrays.asList(condition.getValue().split(",")).contains(String.valueOf(value));
+        return (String.valueOf(contextValue)).indexOf(conditionValue.toString()) > -1;
+
+      case NOT_CONTAINS:
+        if (contextValue != null) {
+          return (String.valueOf(contextValue)).indexOf(conditionValue.toString()) == -1;
         }
         return false;
+
       case REGEX:
-        Pattern pattern = Pattern.compile(condition.getValue());
-        return pattern.matcher(String.valueOf(value)).find();
+        if (contextValue != null) {
+          try {
+            Pattern pattern = Pattern.compile(conditionValue.toString());
+            return pattern.matcher(contextValue.toString()).find();
+          } catch (PatternSyntaxException pse) {
+            return false;
+          }
+        }
+        return false;
+
+      case MODULO:
+        if (contextValue instanceof Number && conditionValue instanceof String) {
+          try {
+            String[] parts = conditionValue.toString().split("\\|");
+            if (parts.length != 2) {
+              return false;
+            }
+            Double divisor = Double.parseDouble(parts[0]);
+            Double remainder = Double.parseDouble(parts[1]);
+            Double value = ((Number) contextValue).doubleValue();
+            return (value % divisor) == remainder;
+          } catch (NumberFormatException nfe) {
+            return false;
+          }
+        }
+        return false;
+
       default:
-        return TypeCasting.compare(operator, value, condition.getValue());
+        if (contextValue == null) {
+          return false;
+        }
+        return TypeCasting.compare(operator, contextValue, conditionValue);
+    }
+  }
+
+  /**
+   * Get context value by property name.
+   *
+   * @param context  Evaluation context.
+   * @param property Property name.
+   * @return Property value.
+   */
+  private static Object getContextValue(EvaluationContext context, String property) {
+    if (property.startsWith("$.")) {
+      return JsonPath
+          .using(jsonPathConfiguration)
+          .parse(mapper.convertValue(context, Map.class))
+          .read(property);
+    }
```

Comment:
What is the returned value in Java in case of error or not found? Just to be sure it doesn't collide with is_set